feature_engine/encoding/count_frequency.py

```diff
@@ -143,22 +175,87 @@ def fit(self, X: pd.DataFrame, y: Optional[pd.Series] = None):
+            for var in self.variables_:
+                temp = {"backoff_bin": 0}
+                for k, v in X[var].value_counts().to_dict().items():
+                    if v <= self.threshold:
+                        temp["backoff_bin"] = temp["backoff_bin"] + v
+                    else:
+                        temp[k] = v
+                self.encoder_dict_[var] = temp
+
+        elif self.threshold is not None and self.encoding_method == "frequency":
+            for var in self.variables_:
+                temp = {"backoff_bin": 0}
+                n_obs = float(len(X))
+                for k, v in (X[var].value_counts() / n_obs).to_dict().items():
+                    if v <= self.threshold:
+                        temp["backoff_bin"] = temp["backoff_bin"] + v
+                    else:
+                        temp[k] = v
+                self.encoder_dict_[var] = temp
+
+        else:
+
             if self.encoding_method == "count":
-                self.encoder_dict_[var] = X[var].value_counts().to_dict()
+                for var in self.variables_:
+                    self.encoder_dict_[var] = X[var].value_counts().to_dict()
 
             elif self.encoding_method == "frequency":
-                n_obs = float(len(X))
-                self.encoder_dict_[var] = (X[var].value_counts() / n_obs).to_dict()
+                for var in self.variables_:
+                    n_obs = float(len(X))
+                    self.encoder_dict_[var] = (X[var].value_counts() / n_obs).to_dict()
 
         self._check_encoding_dictionary()
 
         return self
+
+    def transform(self, X: pd.DataFrame) -> pd.DataFrame:
+        """
+        Replace the categories with counts/frequencies.Replace infrequent categories by
+        the count/frequencies of the 'backoff_bin' based on the threshold provided.
+        In case, no threshold was provided,categories are replaced with
+        counts/frequencies according to encoding_method.
+
+        Parameters
+        ----------
+        X: pandas dataframe of shape = [n_samples, n_features]
+            The input samples.
+
+        Returns
+        -------
+        X: pandas dataframe of shape = [n_samples, n_features]
+            The dataframe where categories are encoded.
+        """
+        X = self._check_transform_input_and_state(X)
+
+        if self.threshold is None:
+            return super().transform(X)
+        else:
+            for feature in self.variables_:
+                keys_before_transform = set(list(X[feature].value_counts().keys()))
+                keys_from_encoder_dict = set(self.encoder_dict_[feature].keys())
+                reminder_keys = keys_before_transform - keys_from_encoder_dict
+                X[feature] = (
+                    X[feature]
+                    .replace(reminder_keys, "backoff_bin")
+                    .map(self.encoder_dict_[feature])
+                )
+            return X
+
+    def inverse_transform(self, X: pd.DataFrame):
```

Comment:
why not?

my idea of the implementation is the following:

category A is present in 10 observations, B in 20 and C in 2 and D in 1.

- if threshold is none, then the encoding_dict = {A:10, B:20, C:2, D:1}
- if threshold is 5, then encoding_dict = {A:10, B:20, C:3, D:3}

Is this what you had in mind? Or what was your idea?

Reply:
Hi,
My idea was, if threshold was None, then return super.transform() as it is as no category swapping is needed, base_encoder's transform works fine.

If some threshold was given, then we have to swap the categories whose count/frequency < threshold with "backoff_bin" category. Since that swapping was missing in the base_encoder transform and introducing NaN's in the transformed DataFrame. I did it this way.

Initially I thought of using base_encoder's transform and replace all the NaN's with backoff bin values, but that would give a warning. Hence did it this way.